src/edu/washington/escience/myria/operator/agg/StatefulUserDefinedAggregator.java

```diff
@@ -0,0 +1,138 @@
+/**
+ *
+ */
+package edu.washington.escience.myria.operator.agg;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.List;
+
+import edu.washington.escience.myria.DbException;
+import edu.washington.escience.myria.Schema;
+import edu.washington.escience.myria.expression.evaluate.GenericEvaluator;
+import edu.washington.escience.myria.expression.evaluate.PythonUDFEvaluator;
+import edu.washington.escience.myria.expression.evaluate.ScriptEvalInterface;
+import edu.washington.escience.myria.storage.AppendableTable;
+import edu.washington.escience.myria.storage.ReadableTable;
+import edu.washington.escience.myria.storage.Tuple;
+import edu.washington.escience.myria.storage.TupleBatch;
+
+/**
+ *
+ */
+public class StatefulUserDefinedAggregator extends UserDefinedAggregator {
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+  /** logger for this class. */
+  private static final org.slf4j.Logger LOGGER =
+      org.slf4j.LoggerFactory.getLogger(StatefulUserDefinedAggregator.class);
+
+  /**
+   *   list for holding state.
+   */
+  private List<TupleBatch> ltb;
+  /**
+   * @param state the initialized state of the tuple
+   * @param updateEvaluator updates the state given an input row
+   * @param pyUDFEvaluators python expression evaluators.
+   * @param emitEvaluators the evaluators that finalize the state
+   * @param resultSchema the schema of the tuples produced by this aggregator
+   */
+  public StatefulUserDefinedAggregator(
+      final Tuple state,
+      final ScriptEvalInterface updateEvaluator,
+      final List<PythonUDFEvaluator> pyUDFEvaluators,
+      final List<GenericEvaluator> emitEvaluators,
+      final Schema resultSchema) {
+    super(state, updateEvaluator, pyUDFEvaluators, emitEvaluators, resultSchema);
+  }
+
+  @Override
+  public void add(final ReadableTable from, final Object state) throws DbException {
+    if (ltb == null) {
+      ltb = new ArrayList<>();
+    }
+    if (!ltb.contains(from)) {
+
+      ltb.add((TupleBatch) from);
+    }
+    try {
+      if (updateEvaluator != null) {
+
+        for (int row = 0; row < from.numTuples(); ++row) {
+          addRow(from, row, state);
+        }
+      }
+
+    } catch (Exception e) {
+
+      throw new DbException("Error updating UDA state", e);
+    }
+  }
+
+  @Override
+  public void add(final List<TupleBatch> from, final Object state) throws DbException {
```

Comment:
I'm confused by the different behaviors of the two `add`s:  one only appends to `ltb` but the other calls `updateEvaluator`. Is it true or false that `updateEvaluator` is null when `pyUDFEvaluators` is set?